Packages/com.chisel.components/Chisel/Components/API.public/BrushMeshAssetFactory/Sphere/BrushFactory.Sphere.cs

```diff
@@ -43,135 +42,28 @@ public static bool GenerateSphereVertices(CSGSphereDefinition definition, ref Ve
 		{
 			definition.Validate();
 			var transform = Matrix4x4.TRS(Vector3.zero, Quaternion.AngleAxis(definition.rotation, Vector3.up), Vector3.one);
-			return GenerateSphereVertices(definition.diameterXYZ, transform, definition.horizontalSegments, definition.verticalSegments, ref vertices);
-		}
-
-		public static bool GenerateSphereVertices(Vector3 diameterXYZ, Matrix4x4 transform, int horzSegments, int vertSegments, ref Vector3[] vertices)
-		{
-			//var lastVertSegment	= vertSegments - 1;
-
-			int vertexCount		= (horzSegments * (vertSegments - 1)) + 2;
-			
-			if (vertices == null ||
-				vertices.Length != vertexCount)
-				vertices = new Vector3[vertexCount];
-			
-			var radius			= 0.5f * diameterXYZ;
-
-			vertices[0] = Vector3.down * radius.y;
-			vertices[1] = Vector3.up   * radius.y;
-			var degreePerSegment	= (360.0f / horzSegments) * Mathf.Deg2Rad;
-			var angleOffset			= ((horzSegments & 1) == 1) ? 0.0f : ((360.0f / horzSegments) * 0.5f) * Mathf.Deg2Rad;
-			for (int v = 1, vertexIndex = 2; v < vertSegments; v++)
-			{
-				var segmentFactor	= ((v - (vertSegments / 2.0f)) / vertSegments);		// [-0.5f ... 0.5f]
-				var segmentDegree	= (segmentFactor * 180);							// [-90 .. 90]
-				var segmentHeight	= Mathf.Sin(segmentDegree * Mathf.Deg2Rad);
-				var segmentRadius	= Mathf.Cos(segmentDegree * Mathf.Deg2Rad);			// [0 .. 0.707 .. 1 .. 0.707 .. 0]
-				for (int h = 0; h < horzSegments; h++, vertexIndex++)
-				{
-					var hRad = (h * degreePerSegment) + angleOffset;
-					vertices[vertexIndex] = new Vector3(Mathf.Cos(hRad) * segmentRadius * radius.x, 
-														segmentHeight                   * radius.y, 
-														Mathf.Sin(hRad) * segmentRadius * radius.z);
-				}
-			}
-			return true;
+			BrushMeshFactory.CreateSphereVertices(definition.diameterXYZ, definition.horizontalSegments, definition.verticalSegments, ref vertices);
+            return true;
 		}
 
 		public static bool GenerateSphereSubMesh(CSGBrushSubMesh subMesh, Vector3 diameterXYZ, Matrix4x4 transform, int horzSegments, int vertSegments, CSGSurfaceAsset[] surfaceAssets, SurfaceDescription[] surfaceDescriptions)
 		{
-			var lastVertSegment	= vertSegments - 1;
-
-			//int vertexCount		= (horzSegments * (vertSegments - 1)) + 2;
-			var triangleCount	= horzSegments + horzSegments;	// top & bottom
-			var quadCount		= horzSegments * (vertSegments - 2);
-			int polygonCount	= triangleCount + quadCount;
-			int halfEdgeCount	= (triangleCount * 3) + (quadCount * 4);
+            var brushMesh = BrushMeshFactory.CreateSphere(diameterXYZ, horzSegments, vertSegments);
 
-			Vector3[] vertices = null;
-			if (!GenerateSphereVertices(diameterXYZ, transform, horzSegments, vertSegments, ref vertices))
-				return false;
+            subMesh.HalfEdges = brushMesh.halfEdges;
+            subMesh.Vertices = brushMesh.vertices;
+            
+            subMesh.Polygons = brushMesh.polygons.Select((p, i) => new CSGBrushSubMesh.Polygon
```

Comment:
Linq is really slow and creates a lot of garbage, we should avoid it if we can. Sure it's fine when you create one Sphere .. but what if you create thousands of spheres from code?